helix-core/src/main/java/org/apache/helix/task/TaskStateModelFactory.java

```diff
@@ -81,6 +79,12 @@ public Thread newThread(Runnable r) {
 
   @Override
   public TaskStateModel createNewStateModel(String resourceName, String partitionKey) {
+    if (_taskExecutor == null) {
```

Comment:
To be safe, add a sync control on this block. There is no guarantee we won't call createNewStateModel in parallel in the future.
And better to put all the logics inside the initializeTaskExecutor().

Reply:
Agreed on the sync part. 

Why do we need all logic inside `initializeTaskExecutor()`?

Reply:
Do you have a scenario that you need to call these methods separately? Put each line of code in separate methods are not efficient. Please try to balance the code.

Reply:
By separating the code into multiple functions based on its purpose, I made it easier to understand. If I combine them, I either name the function `initializeTaskExecutorAndTaskMonitorAndUpdateLiveInstance()` or I force developers to read the code with no clue from the function titles. I'm sure it was clear for you to take a glance and understand what this block is doing because the function names explain them. 

This is a nit, so I don't have that strong if an opinion, but I believe it is correct to break down code into logical groups. To answer your question, "do you have a scenario that you need to call these methods separately?", the answer is yes. I could be utilizing `initializeTaskMonitor()` separately in the other constructor; I'm not doing it because I think that constructor should be gone (see conversation above).

Reply:
It depends, to me the whole init logic is one logical group.
1. each method is small jumping between these methods to read the code is not comfortable.
2. Do you have a scenario that you need to call these methods separately? Or how you support the statement that they are different logical groups?
3. The method name does not need to contain all the logics inside. I don't quite get this point.

Reply:
With the recent changes, `initializeTaskMonitor()` is called by both constructors so it makes sense for it to be factored out. I'm not entirely convinced that it's correct to put all logic in one function - I always thought it's a good practice to break functions down when necessary into logical chunks to avoid big messy functions. 
Since this section of code is outdated, we can discuss in the new conversation if necessary.